### ISSUE TITLE
💅 `noUnusedVariables` misses an arrow function that only references itself

### ISSUE THREAD

**Issue by AlbinoGeek** [biomejs/biome#11950]
### Environment information

<details>

```bash
CLI:
  Version:                      2.5.14

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_DISTRIBUTION:           npm
  JS_RUNTIME_NAME:              node
```

</details>

### Rule name

`correctness/noUnusedVariables`

### Playground link

https://biomejs.dev/playground/?lintRules=noUnusedVariables#files.noUnusedVariables.js=ZQB4AHAAbwByAHQAIABmAHUAbgBjAHQAaQBvAG4AIABmACgAKQAgAHsACgAgACAAZgB1AG4AYwB0AGkAbwBuACAAZwAoACkAIAB7ACAAcgBlAHQAdQByAG4AIABnACgAKQA7ACAAfQAgAC8ALwAgAHIAZQBwAG8AcgB0AGUAZAAKACAAIABsAGUAdAAgAGgAIAA9ACAAKAApACAAPQA%2BACAAaAAoACkAOwAgAC8ALwAgAG4AbwB0ACAAcgBlAHAAbwByAHQAZQBkAAoAIAAgAGMAbwBuAHMAdAAgAGsAIAA9ACAAZgB1AG4AYwB0AGkAbwBuACAAKAApACAAewAgAHIAZQB0AHUAcgBuACAAawAoACkAOwAgAH0AOwAgAC8ALwAgAHIAZQBwAG8AcgB0AGUAZAAKAH0ACgA%3D

```js
export function f() {
  function g() { return g(); } // reported
  let h = () => h(); // not reported
  const k = function () { return k(); }; // reported
}
```

A function declaration or function expression that is only referenced from its own body is reported as unused, but the same pattern with an arrow function is not. ESLint's `no-unused-vars` reports all three.

Reproduced with the 2.5.14 CLI and in the playground (f74eb38).

### Expected result

`h` is reported as unused, like `g` and `k`.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

---

Tool Use Disclaimer: Claude Code using Anthropic Opus 5.5 helped build the reproduction.
